### PR TITLE
Add Slice query modifier

### PR DESCRIPTION
**Pull request by peter-gribanov**
Simplified scheme for dividing the query result into slices.

Before:

```php
$perPage = 20;
$currentPage = 2;

$videos = $rep->match(Spec::andX(
    new VideoPublished(),
    Spec::limit($perPage),
    Spec::offset(($currentPage - 1) * $perPage)
);
```

After:

```php
$perPage = 20;
$currentPage = 2;

$videos = $rep->match(Spec::andX(
    new VideoPublished(),
    Spec::slice($perPage, $currentPage)
);
```

### CODE REVIEW

File: src/Query/Slice.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+namespace Happyr\DoctrineSpecification\Query;
+
+use Happyr\DoctrineSpecification\Logic\AndX;
+
+class Slice extends AndX
```
Comment (Nyholm) [inline src/Query/Slice.php]: Should it really extend AndX?

`Slice::andX` should not be a valid call, right? It is a Specification that **has** an AndX.

Comment (peter-gribanov) [inline src/Query/Slice.php]: I change `Slice` specification

File: src/Query/Slice.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+{
+    /**
+     * @var int
+     */
+    private $sliceSize;
+
+    /**
+     * @var int
+     */
+    private $sliceNumber = 1;
```
Comment (Nyholm) [inline src/Query/Slice.php]: Let's do a zero index. We are programmers and not mathematicians. =)

File: src/Query/Slice.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    /**
+     * @param int $sliceSize
+     * @param int $sliceNumber
+     */
+    public function __construct($sliceSize, $sliceNumber = 1)
```
Comment (Nyholm) [inline src/Query/Slice.php]: Update here

File: src/Query/Slice.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    /**
+     * @param QueryBuilder $qb
+     * @param string       $dqlAlias
+     */
+    public function modify(QueryBuilder $qb, $dqlAlias)
+    {
+        $qb->setMaxResults($this->sliceSize);
+
+        if ($this->sliceNumber > 1) {
```
Comment (Nyholm) [inline src/Query/Slice.php]: and here.

File: src/Query/Slice.php (hunk 2 above)
Comment (peter-gribanov) [inline src/Query/Slice.php]: If we change the index, then the behavior changes too.

```php
$perPage = 20;
$currentPage = 2;

$videos = $rep->match(Spec::andX(
    new VideoPublished(),
    Spec::slice($perPage, $currentPage - 1)
);
```
The first page, with pagination, will be with index 1.
This is the whole point of this change.

For comparison, look at what led to the change of the index when pagination in [Yii](https://github.com/yiisoft/yii/blob/1.1.17/framework/web/CPagination.php).
The current page in the request is not equal to the current page in the paginator.

Comment (Nyholm) [inline src/Query/Slice.php]: > The first page, with pagination, will be with index 1.

Why?

The first item in the list should always be zero. In computer science we zero index everything.

Comment (peter-gribanov) [inline src/Query/Slice.php]: > In computer science we zero index everything.

Certainly. But business logic is focused on the real world:
Get first slice = get slice number one = `$sliceNumber = 1`

In the specifications we focus on business logic.

Example:
https://github.com/Happyr/Doctrine-Specification/issues?page=2&q=is%3Aissue+is%3Aclosed
* `page=2` - in real we select page number 2. In business logic we use number 2, but in persistence level we use number 1 *(current page - 1)*.

Therefore, i propose to divide the business layer and the persistence layer and hide the offset in the specification.

Comment (Nyholm) [inline src/Query/Slice.php]: I agree that in the URL, the first page should be 1. Just like GitHub does. But this is as far from business logic we can get. It is about a database call. 

Ill be happy to merge if you change it two zero indexed and avoid all confusion.

**Comment by peter-gribanov** [comment]
* change `$sliceNumber` to `0`
* add tests

**Pull request merged**